### ISSUE TITLE
[apostrophe-modals] Dismissing modals by clicking outside of them dangerous with unsaved edits in those modals

### ISSUE THREAD

**Issue by stuartromanek** [punkave/apostrophe#503]
Needs a confirmation or everything needs to rely on the clicking of a Save or Cancel type action button

**Comment by boutell** [comment]
We fixed this in 0.5, so I assume this is an 0.6 ticket. In 0.5 we use a
`beforeCancel` method in editor.js of `apostrophe-snippets`:

```
        beforeCancel: function(callback) {
          if (confirm('Are you sure you want to discard this ' +
options.instanceLabel.toLowerCase() + '?')) {
            return callback(null);
          }
          return callback('notconfirmed');
        },
```

`beforeCancel` also exists in `apostrophe-modal` in 0.6 so it should be
easy to use the same mechanism. Or, hey, a confirm prompt that isn't a
crappy 1995 confirm() call... like, our own async confirm modal... someday
we should do that... but, separate ticket (:


On Thu, Apr 14, 2016 at 9:57 AM, Stuart Romanek <notifications@github.com>
wrote:

> Needs a confirmation or everything needs to rely on the clicking of a Save
> or Cancel type action button
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/punkave/apostrophe/issues/503>
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com

**Comment by jth-** [comment]
I've implemented this rudimentarily here: 9140cfc37aa10ceb4dd0484805295b85ea8f6eec

I've got some comments in that particular commit message, let me know if that seems like a good way of doing this @boutell

**Comment by boutell** [comment]
Hmm... better than what we have in 0.5, because there's some generality.
But I would further suggest instead implementing an unsavedChanges()
method, which by default returns false. This gives modals the option of
only checking for unsaved changes when they actually need to decide whether
to allow a cancel, which is probably cheaper than change() handlers on
everything under the sun.

Also, I hate to say this, but unsavedChanges() should support at least an
optional callback. One of the best ways of implementing it for many modals
would be to call convert() and compare the result to the original object,
and convert() is async.

On Fri, Apr 15, 2016 at 4:12 PM, Jimmy Harrison <notifications@github.com>
wrote:

> I've implemented this rudimentarily here: 9140cfc
> <https://github.com/punkave/apostrophe/commit/9140cfc37aa10ceb4dd0484805295b85ea8f6eec>
>
> I've got some comments in that particular commit message, let me know if
> that seems like a good way of doing this @boutell
> <https://github.com/boutell>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/punkave/apostrophe/issues/503#issuecomment-210622810>
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com

**Comment by boutell** [comment]
However what you have here is a step in the right direction and entirely
acceptable for this week if that's a concern!

On Fri, Apr 15, 2016 at 11:36 PM, Tom Boutell <tom@punkave.com> wrote:

> Hmm... better than what we have in 0.5, because there's some generality.
> But I would further suggest instead implementing an unsavedChanges()
> method, which by default returns false. This gives modals the option of
> only checking for unsaved changes when they actually need to decide whether
> to allow a cancel, which is probably cheaper than change() handlers on
> everything under the sun.
>
> Also, I hate to say this, but unsavedChanges() should support at least an
> optional callback. One of the best ways of implementing it for many modals
> would be to call convert() and compare the result to the original object,
> and convert() is async.
>
> On Fri, Apr 15, 2016 at 4:12 PM, Jimmy Harrison <notifications@github.com>
> wrote:
>
>> I've implemented this rudimentarily here: 9140cfc
>> <https://github.com/punkave/apostrophe/commit/9140cfc37aa10ceb4dd0484805295b85ea8f6eec>
>>
>> I've got some comments in that particular commit message, let me know if
>> that seems like a good way of doing this @boutell
>> <https://github.com/boutell>
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/punkave/apostrophe/issues/503#issuecomment-210622810>
>>
>
>
>
> --
>
>
> *THOMAS BOUTELL, *DEV & OPS
> P'UNK AVENUE | (215) 755-1330  |  punkave.com
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com

**Comment by boutell** [comment]
The more I think about it the more it probably makes sense for field types
to implement a 'change' event of their own if they aren't just native
controls that already have one, and for modals to catch all 'change' events
that bubble up to them and use them to set your flag. No need for async
jazz.

On Fri, Apr 15, 2016 at 11:36 PM, Tom Boutell <tom@punkave.com> wrote:

> However what you have here is a step in the right direction and entirely
> acceptable for this week if that's a concern!
>
> On Fri, Apr 15, 2016 at 11:36 PM, Tom Boutell <tom@punkave.com> wrote:
>
>> Hmm... better than what we have in 0.5, because there's some generality.
>> But I would further suggest instead implementing an unsavedChanges()
>> method, which by default returns false. This gives modals the option of
>> only checking for unsaved changes when they actually need to decide whether
>> to allow a cancel, which is probably cheaper than change() handlers on
>> everything under the sun.
>>
>> Also, I hate to say this, but unsavedChanges() should support at least an
>> optional callback. One of the best ways of implementing it for many modals
>> would be to call convert() and compare the result to the original object,
>> and convert() is async.
>>
>> On Fri, Apr 15, 2016 at 4:12 PM, Jimmy Harrison <notifications@github.com
>> > wrote:
>>
>>> I've implemented this rudimentarily here: 9140cfc
>>> <https://github.com/punkave/apostrophe/commit/9140cfc37aa10ceb4dd0484805295b85ea8f6eec>
>>>
>>> I've got some comments in that particular commit message, let me know if
>>> that seems like a good way of doing this @boutell
>>> <https://github.com/boutell>
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/punkave/apostrophe/issues/503#issuecomment-210622810>
>>>
>>
>>
>>
>> --
>>
>>
>> *THOMAS BOUTELL, *DEV & OPS
>> P'UNK AVENUE | (215) 755-1330  |  punkave.com
>>
>
>
>
> --
>
>
> *THOMAS BOUTELL, *DEV & OPS
> P'UNK AVENUE | (215) 755-1330  |  punkave.com
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com
